### ISSUE TITLE
Very slow - needs cache

### ISSUE THREAD

**Issue by ollyde** [westerlind/alfred-raindrop-search#13]
I was traveling the last few months; using this plugin.

I found that sometimes it would take 5-10 seconds to load bookmarks.

At home now on fiber and sometimes it takes over 3 seconds when typing.

This makes the Alfred plugin useless, consider adding cache here.

**Comment by westerlind** [comment]
@ollydixon 

I am working on a feature for this, but it's not done yet, and I'm not sure exactly when it will be.

What it will do (at least as it looks at the moment) is that it adds a third entry point to the workflow that does search in an entirely different way than the current search feature.
So right now there is the server side search, the bookmark adding feature, and there will in the future also be a cached search feature.
You will be able to simply use this search variant instead of the old one if you want to, or use them both with different keywords (like r for search now) or shortcuts.

The difference between the server side search and the cached search is that with the server side search it will be a bit slower (like it always is now), but you get for example full text search, more advanced logic to finding things that match what you searched for, and the search includes newly added bookmarks right away

The cached search will update it's cache sometimes in the background, and then use only the cache for searching, so it will take a bit longer for new bookmarks to be included, and it will not have full text search or the better search algorithm that is available on the server side etc, but it will show the results almost instantly and it will also (which might end up as an option) let Alfred have a say in the order of the results list so that more commonly searched for items will be further up in the results list. (this breaks things in the classic search variant, which is why it doesn't do that)

Quite a bit of this is already done, and all the concepts are tested, but I'm not sure when this will be ready for release.
Probably within a few months.

**Comment by ollyde** [comment]
@westerlind awesome thanks for the update; looking forwards to test it! Let me know what that is :-)

**Comment by Laptop765** [comment]
@westerlind I have a [rough proof of concept](https://github.com/westerlind/alfred-raindrop-search/compare/master...Laptop765:alfred-raindrop-search:master) illustrating the use of the [`httpcache`](https://github.com/gregjones/httpcache) library for `go`. It works by leveraging the [`etag` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) returned by Raindrop's API to enable reliable client-side caching.

It provides an indeterminate speedup on my machine _when the same query is issued_ but given the way things currently work in this codebase each query is a different request so there may be limited value. One way of leveraging it would be to download everything locally through the cache and search that way but that would probably require some testing on how long the cache lasts since it would be counter-productive to redownload everything for every search 😅 That would also lose server-side querying and might impact functionality?

So to summarize: I'm not sure this is useful but wanted to share my findings just in case.

**Comment by ghost** [comment]
Hi @westerlind, thanks so much for this workflow! Was curious as to whether the cache was ever implemented 🙏

**Comment by westerlind** [comment]
It's not, unfortunately. But I still intend to do it, I just haven't gotten to it yet.

**Comment by ghost** [comment]
OK, look forward to it 😄

**Comment by ollyde** [comment]
Switched to Raycast now; way better than Alfred!

**Comment by byalextran** [comment]
+1 for caching. i just switched to raindrop from pinboard and was using [a pinboard alfred workflow](https://github.com/spamwax/alfred-pinboard-rs) that had caching. 

coming from that to this workflow, the lack of caching is noticeable. particularly since i'm assuming all us alfred users are hyper productive and efficient people. a second delay searching for bookmarks can feel like an eternity. 🤣

perhaps there's a setting that could automatically update the cache after X period and/or provide a command to manually update the cache. but the default is that searches are done from a cache.

**Comment by westerlind** [comment]
What I intend to do is to eventually provide both ways of searching, and then you could for example set "r" to the one you prefer, and same with any shortcut you might be using.

It's all about how you are using this workflow and what you are using bookmarks for, and for me it's more important with deeper and more accurate search than being a second or two faster at each search, which is why it works like it does currently. The current version does make it faster in practice for me personally for that reason.

Using a cache will make each search much faster, but will also prevent it from searching the full text of a bookmarked page, which will lead to the results not being as good, which will mean that it's more likely that you will have to make more searches to find what you are looking for, and in that way it might actually be slower depending on how you are using it.

But I'm using bookmarks mainly as some kind of knowledge database, and I understand that if you use it as a fast way to open a set of webpages you are visiting somewhat often, that makes the current behavior feel slow in a way that it doesn't if you use it like I do.

**Comment by byalextran** [comment]
gotcha, yeah, it sounds like my workflow is different than yours. i don't need full-text search. i search strictly off the bookmark title. so it would be helpful if you offered both ways of searching. thanks!

**Comment by westerlind** [comment]
This feature is now finally implemented, and available in version 3.0, which was released a moment ago!
See the release notes here: https://github.com/westerlind/alfred-raindrop-search/releases/tag/v3.0
And there is also more information about how it functions in the readme, here: https://github.com/westerlind/alfred-raindrop-search
